File: Ruby/CocktailSort/stack_array_cocktail_sort.rb
require 'benchmark'
require 'objspace'

class Rating
  attr_accessor :user_id, :movie_id, :rating, :timestamp
  def initialize(user_id, movie_id, rating, timestamp)
    @user_id = user_id
    @movie_id = movie_id
    @rating = rating
    @timestamp = timestamp
  end
  def <=>(other)
    @rating <=> other.rating
  end
  def to_s
    "#{@user_id},#{@movie_id},#{@rating},#{@timestamp}"
  end
end

class StackArrayCocktailSort
  attr_accessor :stack, :top
  def initialize(capacity)
    @stack = []
    @top = -1
  end
  def push(rating)
    @stack << rating
    @top += 1
  end
  def cocktail_sort
    swapped = true
    start_i = 0
    end_i = @top
    while swapped
      swapped = false
      (start_i...end_i).each do |i|
        if @stack[i] > @stack[i + 1]
          @stack[i], @stack[i + 1] = @stack[i + 1], @stack[i]
          swapped = true
        end
      end
      break unless swapped
      swapped = false
      end_i -= 1
      end_i.downto(start_i + 1) do |i|
        if @stack[i - 1] > @stack[i]
          @stack[i], @stack[i - 1] = @stack[i - 1], @stack[i]
          swapped = true
        end
      end
      start_i += 1
    end
  end
end

if ARGV.size < 2
  puts "Uso: ruby stack_array_cocktail_sort.rb <arquivo.csv> <N>"
  exit
end

path = ARGV[0]
n = ARGV[1].to_i
pilha = StackArrayCocktailSort.new(n)

File.open(path) do |f|
  f.readline
  f.each_line.take(n).each do |line|
    user_id, movie_id, rating, timestamp = line.strip.split(',')
    pilha.push(Rating.new(user_id.to_i, movie_id.to_i, rating.to_f, timestamp.to_i))
  end
end

mem_before = ObjectSpace.memsize_of_all
time = Benchmark.realtime { pilha.cocktail_sort }
mem_after = ObjectSpace.memsize_of_all

puts "Ordenado #{n} registros em #{(time * 1000).round(2)} ms"
puts "Memória usada: #{((mem_after - mem_before) / 1024.0).round(2)} KB"
puts "Primeiros 100 registros ordenados:"
pilha.stack.first(100).each { |r| puts r }

puts "Últimos 100 registros ordenados:"
pilha.stack.last(100).each { |r| puts r }

File: JAVA/Cocktail Sort/StackArrayCocktailSort.java
import java.io.*;

public class StackArrayCocktailSort
{
    static class Rating
    {
        int userId, movieId;
        float rating;
        long timestamp;

        Rating(int u, int m, float r, long t)
        {
            userId = u; movieId = m; rating = r; timestamp = t;
        }

        int compareTo(Rating o)
        {
            return Float.compare(this.rating, o.rating);
        }

        public String toString()
        {
            return userId + "," + movieId + "," + rating + "," + timestamp;
        }
    }

    Rating[] stack;
    int top = -1;

    StackArrayCocktailSort(int capacity)
    {
        stack = new Rating[capacity];
    }

    void push(Rating r)
    {
        stack[++top] = r;
    }

    void cocktailSort()
    {
        boolean swapped = true;
        int start = 0, end = top;

        while (swapped)
        {
            swapped = false;
            for (int i = start; i < end; i++)
            {
                if (stack[i].compareTo(stack[i + 1]) > 0) 
                {
                    Rating temp = stack[i];
                    stack[i] = stack[i + 1];
                    stack[i + 1] = temp;
                    swapped = true;
                }
            }

            if (!swapped) break;
            swapped = false;
            end--;

            for (int i = end; i > start; i--)
            {
                if (stack[i - 1].compareTo(stack[i]) > 0)
                {
                    Rating temp = stack[i];
                    stack[i] = stack[i - 1];
                    stack[i - 1] = temp;
                    swapped = true;
                }
            }
            start++;
        }
    }

    public static void main(String[] args)
    throws Exception
    {
        if (args.length < 2)
        {
            System.out.println("Uso: java StackArrayCocktailSort <ratings.csv> <N>");
            return;
        }

        String path = args[0];
        int N = Integer.parseInt(args[1]);
        StackArrayCocktailSort pilha = new StackArrayCocktailSort(N);

        try (BufferedReader br = new BufferedReader(new FileReader(path)))
        {
            String line = br.readLine(); int count = 0;
            while ((line = br.readLine()) != null && count < N)
            {
                String[] f = line.split(",");
                pilha.push(new Rating(Integer.parseInt(f[0]), Integer.parseInt(f[1]),
                                      Float.parseFloat(f[2]), Long.parseLong(f[3])));
                count++;
            }
        }

        long t0 = System.nanoTime();
        pilha.cocktailSort();
        long t1 = System.nanoTime();

        System.out.printf("Ordenado %d registros em %.2f ms\n", N, (t1 - t0) / 1e6);
        for (int i = 0; i <= Math.min(100, pilha.top); i++)
        {
            System.out.println(pilha.stack[i]);
        }
    }
}
